File: src/mentoai/ops.py
"""CLI와 어드민 API가 공유하는 운영 작업 계층.

한 곳에 구현해 두 인터페이스가 항상 같은 로직을 돌린다.
"""

import asyncio
import logging
from typing import Any

from pydantic import BaseModel, Field

from mentoai.ai.embeddings import model_key
from mentoai.config import get_settings
from mentoai.db.pool import execute, fetch, fetchrow

logger = logging.getLogger(__name__)


# ---------- 백그라운드 작업 가드 ----------

_running: set[str] = set()


def start_background(name: str, factory) -> bool:
    """factory: 인자 없는 코루틴 팩토리. 동일 이름 작업이 이미 돌면 False.

    팩토리를 받는 이유: 가드에 걸린 호출에서 코루틴이 생성만 되고
    await되지 않는 'never awaited' 경로를 원천 차단한다.
    """
    if name in _running:
        return False
    task = asyncio.create_task(factory())
    _running.add(name)
    task.add_done_callback(lambda _: _running.discard(name))
    return True


def running_ops() -> list[str]:
    return sorted(_running)


# ---------- 현황 ----------

async def get_status() -> dict[str, Any]:
    settings = get_settings()
    counts = await fetchrow(
        """
        SELECT
          (SELECT count(*) FROM bronze.raw_postings) AS bronze,
          (SELECT count(*) FROM silver.jobs) AS jobs,
          (SELECT count(*) FROM silver.job_embeddings) AS embeddings,
          (SELECT count(*) FROM users) AS users,
          (SELECT count(*) FROM analysis_cache) AS cached_analyses
        """
    )
    assert counts is not None
    sizes = await fetchrow(
        """
        SELECT
          pg_size_pretty(pg_total_relation_size('bronze.raw_postings'::regclass)) AS bronze_size,
          pg_size_pretty(pg_total_relation_size('silver.jobs'::regclass)) AS jobs_size,
          pg_size_pretty(pg_total_relation_size('silver.job_embeddings'::regclass)) AS embeddings_size
        """
    )
    last_run = await fetchrow(
        """
        SELECT id, status, started_at, finished_at, scraped, silver_upserted, embedded
        FROM pipeline_runs ORDER BY id DESC LIMIT 1
        """
    )

    from mentoai.api.scheduler import schedule_info

    return {
        **dict(counts),
        "sizes": dict(sizes) if sizes else {},
        "embedding_model": model_key(),
        "embedding_dim": settings.embedding_dim,
        "gemini_model": settings.gemini_model,
        "schedule": schedule_info(),
        "running": running_ops(),
        "last_run": dict(last_run) if last_run else None,
    }


async def list_runs(limit: int = 20) -> list[dict[str, Any]]:
    limit = min(max(limit, 1), 100)
    rows = await fetch(
        """
        SELECT id, status, started_at, finished_at, scraped, silver_upserted, embedded, error
        FROM pipeline_runs ORDER BY id DESC LIMIT $1
        """,
        limit,
    )
    return [dict(r) for r in rows]


# ---------- 공고 ----------

async def list_jobs(query: str = "", limit: int = 30) -> list[dict[str, Any]]:
    limit = min(max(limit, 1), 100)
    if query:
        rows = await fetch(
            """
            SELECT id, source, source_id, company, position, skill_tags, updated_at
            FROM silver.jobs
            WHERE company ILIKE $1 OR position ILIKE $1
            ORDER BY updated_at DESC LIMIT $2
            """,
            f"%{query}%",
            limit,
        )
    else:
        rows = await fetch(
            """
            SELECT id, source, source_id, company, position, skill_tags, updated_at
            FROM silver.jobs ORDER BY updated_at DESC LIMIT $1
            """,
            limit,
        )
    return [dict(r) for r in rows]


async def delete_job(job_id: int) -> dict[str, Any]:
    """silver 공고와 대응하는 bronze 원본까지 지운다 (임베딩/캐시는 cascade)."""
    job = await fetchrow(
        "SELECT source, source_id, company, position FROM silver.jobs WHERE id = $1", job_id
    )
    if not job:
        raise KeyError(f"공고 없음: {job_id}")
    await execute("DELETE FROM silver.jobs WHERE id = $1", job_id)
    await execute(
        "DELETE FROM bronze.raw_postings WHERE source = $1 AND source_id = $2",
        job["source"],
        job["source_id"],
    )
    logger.info("공고 삭제: #%s %s %s", job_id, job["company"], job["position"])
    return dict(job)


# ---------- 인재 ----------

class UserPayload(BaseModel):
    username: str = Field(min_length=1, max_length=30)
    desired_job: str = Field(min_length=1, max_length=60)
    career_years: int = Field(ge=0, le=40)
    skills: list[str] = Field(default_factory=list, max_length=20)


async def create_user(payload: UserPayload) -> dict[str, Any]:
    row = await fetchrow(
        "INSERT INTO users (username) VALUES ($1) ON CONFLICT (username) DO NOTHING RETURNING id",
        payload.username,
    )
    if row is None:
        raise ValueError(f"이미 존재하는 사용자: {payload.username}")
    await _upsert_spec(row["id"], payload)
    return {"id": row["id"], **payload.model_dump()}


async def update_user(user_id: int, payload: UserPayload) -> dict[str, Any]:
    row = await fetchrow("SELECT id, username FROM users WHERE id = $1", user_id)
    if not row:
        raise KeyError(f"사용자 없음: {user_id}")
    await _upsert_spec(user_id, payload)
    return {"id": user_id, **payload.model_dump()}


async def _upsert_spec(user_id: int, payload: UserPayload) -> None:
    await execute(
        """
        INSERT INTO user_specs (user_id, desired_job, career_years, skills)
        VALUES ($1, $2, $3, $4)
        ON CONFLICT (user_id) DO UPDATE SET
            desired_job = EXCLUDED.desired_job,
            career_years = EXCLUDED.career_years,
            skills = EXCLUDED.skills
        """,
        user_id,
        payload.desired_job,
        payload.career_years,
        payload.skills,
    )


async def find_user_by_name(username: str) -> dict[str, Any] | None:
    row = await fetchrow("SELECT id, username FROM users WHERE username = $1", username)
    return dict(row) if row else None


async def delete_user(user_id: int) -> None:
    result = await execute("DELETE FROM users WHERE id = $1", user_id)
    if result == "DELETE 0":
        raise KeyError(f"사용자 없음: {user_id}")


# ---------- 임베딩 ----------

async def rebuild_embeddings() -> int:
    """전량 재계산: 임베딩은 파생 데이터라 전체 삭제 후 재적재한다."""
    await execute("DELETE FROM silver.job_embeddings")
    from mentoai.pipeline import gold

    return await gold.run()


def embedding_models() -> dict[str, Any]:
    from fastembed import TextEmbedding

    return {
        "current": model_key(),
        "provider": get_settings().embedding_provider,
        "dim": get_settings().embedding_dim,
        "fastembed": sorted(
            (
                {"model": m["model"], "dim": m["dim"], "size_gb": m["size_in_GB"]}
                for m in TextEmbedding.list_supported_models()
            ),
            key=lambda m: m["model"],
        ),
        "gemini_default": "gemini-embedding-001",
    }


async def switch_embedding_model(provider: str, model: str | None) -> dict[str, Any]:
    from mentoai.embed_switch import switch_embedding

    return await switch_embedding(provider=provider, model=model)


# ---------- 캐시 ----------

async def list_cache() -> list[dict[str, Any]]:
    rows = await fetch(
        """
        SELECT c.job_id, c.user_id, c.model, c.created_at,
               u.username, j.company, j.position
        FROM analysis_cache c
        JOIN users u ON u.id = c.user_id
        JOIN silver.jobs j ON j.id = c.job_id
        ORDER BY c.created_at DESC LIMIT 20
        """
    )
    return [dict(r) for r in rows]


async def clear_cache() -> int:
    rows = await fetch("DELETE FROM analysis_cache RETURNING job_id")
    return len(rows)


async def delete_cache(job_id: int, user_id: int) -> None:
    result = await execute(
        "DELETE FROM analysis_cache WHERE job_id = $1 AND user_id = $2", job_id, user_id
    )
    if result == "DELETE 0":
        raise KeyError(f"캐시 없음: job={job_id} user={user_id}")

File: internal/ops/ops.go
// Package ops는 CLI와 어드민 API가 공유하는 운영 작업 계층이다.
// 한 곳에 구현해 두 인터페이스가 항상 같은 로직을 돌린다.
package ops

import (
	"context"
	"fmt"
	"log/slog"
	"os"
	"sort"
	"strconv"
	"sync"

	"github.com/Chae-JS/mentoai/internal/config"
	"github.com/Chae-JS/mentoai/internal/domain"
	"github.com/Chae-JS/mentoai/internal/embedding"
	"github.com/Chae-JS/mentoai/internal/envfile"
	"github.com/Chae-JS/mentoai/internal/storage"
)

// PipelineRunner는 파이프라인 실행 어댑터다 (composition root에서 조립).
type PipelineRunner interface {
	// RunPipeline은 bronze→silver→gold 전체를 실행한다.
	RunPipeline(ctx context.Context) (domain.PipelineResult, error)
	// EmbedAll은 gold(임베딩)만 실행한다.
	EmbedAll(ctx context.Context) (int, error)
}

// ScheduleInfoProvider는 스케줄 상태를 돌려준다 (스케줄러가 구현).
type ScheduleInfoProvider interface {
	Info() domain.ScheduleInfo
}

// Service는 운영 서비스다.
type Service struct {
	cfg      *config.Holder
	users    storage.UserRepo
	raw      storage.RawPostingRepo
	jobs     storage.JobRepo
	embeds   storage.EmbeddingRepo
	cache    storage.CacheRepo
	runs     storage.RunRepo
	probe    storage.SizeProbe
	registry *embedding.Registry
	active   *embedding.AtomicEmbedder
	runner   PipelineRunner
	schedule ScheduleInfoProvider
	guard    *Guard
}

// New는 의존성을 주입받아 운영 서비스를 만든다.
func New(
	cfg *config.Holder,
	users storage.UserRepo,
	raw storage.RawPostingRepo,
	jobs storage.JobRepo,
	embeds storage.EmbeddingRepo,
	cache storage.CacheRepo,
	runs storage.RunRepo,
	probe storage.SizeProbe,
	registry *embedding.Registry,
	active *embedding.AtomicEmbedder,
	runner PipelineRunner,
	schedule ScheduleInfoProvider,
	reporter ErrorReporter,
) *Service {
	return &Service{
		cfg: cfg, users: users, raw: raw, jobs: jobs, embeds: embeds,
		cache: cache, runs: runs, probe: probe, registry: registry,
		active: active, runner: runner, schedule: schedule, guard: NewGuard(reporter),
	}
}

// ---------- 백그라운드 작업 가드 ----------

// ErrorReporter는 에러 리포팅 포트다 (telemetry.Reporter가 구조적으로 충족).
type ErrorReporter interface {
	CaptureError(err error, tags map[string]string)
}

// Guard는 이름 기반 단일 실행 가드다. Python의 start_background와 동일한 의미다.
type Guard struct {
	mu       sync.Mutex
	running  map[string]struct{}
	reporter ErrorReporter
}

func NewGuard(reporter ErrorReporter) *Guard {
	return &Guard{running: map[string]struct{}{}, reporter: reporter}
}

// Start는 동일 이름 작업이 이미 돌면 false를 반환하고, 아니면 고루틴으로 실행한다.
func (g *Guard) Start(name string, fn func(ctx context.Context) error) bool {
	g.mu.Lock()
	if _, dup := g.running[name]; dup {
		g.mu.Unlock()
		return false
	}
	g.running[name] = struct{}{}
	g.mu.Unlock()

	go func() {
		defer func() {
			if r := recover(); r != nil {
				slog.Error("백그라운드 작업 패닉", "op", name, "panic", r)
				if g.reporter != nil {
					g.reporter.CaptureError(fmt.Errorf("백그라운드 작업 패닉 (%s): %v", name, r),
						map[string]string{"op": name, "kind": "panic"})
				}
			}
			g.mu.Lock()
			delete(g.running, name)
			g.mu.Unlock()
		}()
		if err := fn(context.Background()); err != nil {
			slog.Error("백그라운드 작업 실패", "op", name, "error", err)
			if g.reporter != nil {
				g.reporter.CaptureError(err, map[string]string{"op": name})
			}
		}
	}()
	return true
}

// Running은 실행 중 작업 이름을 정렬해 돌려준다.
func (g *Guard) Running() []string {
	g.mu.Lock()
	defer g.mu.Unlock()
	out := make([]string, 0, len(g.running))
	for name := range g.running {
		out = append(out, name)
	}
	sort.Strings(out)
	return out
}

// ---------- 현황 ----------

// Status는 대시보드 현황을 집계한다.
func (s *Service) Status(ctx context.Context) (domain.Status, error) {
	var out domain.Status
	bronze, err := s.raw.Count(ctx)
	if err != nil {
		return out, err
	}
	jobs, err := s.jobs.Count(ctx)
	if err != nil {
		return out, err
	}
	embeds, err := s.embeds.Count(ctx)
	if err != nil {
		return out, err
	}
	users, err := s.users.Count(ctx)
	if err != nil {
		return out, err
	}
	cached, err := s.cache.Count(ctx)
	if err != nil {
		return out, err
	}
	lastRun, err := s.runs.Latest(ctx)
	if err != nil {
		return out, err
	}

	settings := s.cfg.Get()
	out = domain.Status{
		Bronze:         bronze,
		Jobs:           jobs,
		Embeddings:     embeds,
		Users:          users,
		CachedAnalyses: cached,
		Sizes:          s.sizes(ctx),
		EmbeddingModel: s.active.Current().ModelKey(),
		EmbeddingDim:   settings.EmbeddingDim,
		GeminiModel:    settings.GeminiModel,
		Schedule:       s.schedule.Info(),
		Running:        s.guard.Running(),
		LastRun:        lastRun,
	}
	if out.Running == nil {
		out.Running = []string{}
	}
	return out, nil
}

// sizes는 테이블 용량을 사람이 읽는 형태로 조회한다. dbstat 미지원 시 "-"다.
func (s *Service) sizes(ctx context.Context) domain.Sizes {
	out := domain.Sizes{BronzeSize: "-", JobsSize: "-", EmbeddingsSize: "-"}
	bronze, jobs, embeds, err := s.probe.TableSizes(ctx)
	if err != nil {
		return out
	}
	out.BronzeSize = prettyBytes(bronze)
	out.JobsSize = prettyBytes(jobs)
	out.EmbeddingsSize = prettyBytes(embeds)
	return out
}

// prettyBytes는 pg_size_pretty 비슷한 표기를 만든다.
func prettyBytes(n int64) string {
	const unit = 1024
	if n < unit {
		return fmt.Sprintf("%d bytes", n)
	}
	value := float64(n)
	for _, suffix := range []string{"kB", "MB", "GB", "TB"} {
		value /= unit
		if value < unit {
			return fmt.Sprintf("%.1f %s", value, suffix)
		}
	}
	return fmt.Sprintf("%.1f PB", value/unit)
}

// ListRuns는 최근 파이프라인 실행 이력을 돌려준다.
func (s *Service) ListRuns(ctx context.Context, limit int) ([]domain.RunRow, error) {
	rows, err := s.runs.List(ctx, clamp(limit, 20))
	return rows, err
}

// ---------- 공고 ----------

// ListJobs는 회사명·포지션 검색으로 공고 목록을 돌려준다.
func (s *Service) ListJobs(ctx context.Context, query string, limit int) ([]domain.JobAdminRow, error) {
	return s.jobs.List(ctx, query, clamp(limit, 30))
}

// DeleteJob은 silver 공고와 대응하는 bronze 원본까지 지운다 (임베딩/캐시는 cascade).
func (s *Service) DeleteJob(ctx context.Context, jobID int64) error {
	key, err := s.jobs.GetKey(ctx, jobID)
	if err != nil {
		return err
	}
	if key == nil {
		return domain.NotFound(fmt.Sprintf("공고 없음: %d", jobID))
	}
	if _, err := s.jobs.Delete(ctx, jobID); err != nil {
		return err
	}
	if _, err := s.raw.DeleteByKey(ctx, key.Source, key.SourceID); err != nil {
		return err
	}
	// cascade로 사라진 임베딩 행이 검색 인덱스 캐시에 남지 않게 한다.
	s.embeds.Invalidate(ctx)
	slog.Info("공고 삭제", "id", jobID)
	return nil
}

// ---------- 인재 ----------

// CreateUser는 사용자와 스펙을 등록한다. 중복 이름은 409 오류다.
func (s *Service) CreateUser(ctx context.Context, p domain.UserPayload) (domain.UserResponse, error) {
	var out domain.UserResponse
	if err := p.Validate(); err != nil {
		return out, domain.Unprocessable(err.Error())
	}
	id, err := s.users.Insert(ctx, p.Username)
	if err != nil {
		return out, err
	}
	if id == 0 {
		return out, domain.Conflict(fmt.Sprintf("이미 존재하는 사용자: %s", p.Username))
	}
	if err := s.users.UpsertSpec(ctx, id, p.DesiredJob, p.CareerYears, p.Skills); err != nil {
		return out, err
	}
	return toUserResponse(id, p), nil
}

// UpdateUser는 사용자 스펙을 갱신한다. 사용자가 없으면 404다.
func (s *Service) UpdateUser(ctx context.Context, userID int64, p domain.UserPayload) (domain.UserResponse, error) {
	var out domain.UserResponse
	if err := p.Validate(); err != nil {
		return out, domain.Unprocessable(err.Error())
	}
	row, err := s.users.GetByID(ctx, userID)
	if err != nil {
		return out, err
	}
	if row == nil {
		return out, domain.NotFound(fmt.Sprintf("사용자 없음: %d", userID))
	}
	if err := s.users.UpsertSpec(ctx, userID, p.DesiredJob, p.CareerYears, p.Skills); err != nil {
		return out, err
	}
	return toUserResponse(userID, p), nil
}

func toUserResponse(id int64, p domain.UserPayload) domain.UserResponse {
	skills := p.Skills
	if skills == nil {
		skills = []string{}
	}
	return domain.UserResponse{
		ID: id, Username: p.Username, DesiredJob: p.DesiredJob,
		CareerYears: p.CareerYears, Skills: skills,
	}
}

// FindUserByName은 이름으로 사용자를 찾는다.
func (s *Service) FindUserByName(ctx context.Context, username string) (*domain.UserRow, error) {
	return s.users.FindByUsername(ctx, username)
}

// DeleteUser는 사용자를 삭제한다(관련 캐시 cascade).
func (s *Service) DeleteUser(ctx context.Context, userID int64) error {
	affected, err := s.users.Delete(ctx, userID)
	if err != nil {
		return err
	}
	if affected == 0 {
		return domain.NotFound(fmt.Sprintf("사용자 없음: %d", userID))
	}
	return nil
}

// ---------- 임베딩 ----------

// RebuildEmbeddings는 전량 재계산한다. 임베딩은 파생 데이터라 전체 삭제 후 재적재한다.
func (s *Service) RebuildEmbeddings(ctx context.Context) (int, error) {
	if err := s.embeds.DeleteAll(ctx); err != nil {
		return 0, err
	}
	return s.runner.EmbedAll(ctx)
}

// EmbeddingModels는 현재 선택과 등록된 공급자 목록을 돌려준다.
func (s *Service) EmbeddingModels() domain.EmbeddingModels {
	settings := s.cfg.Get()
	return domain.EmbeddingModels{
		Current:       s.active.Current().ModelKey(),
		Provider:      settings.EmbeddingProvider,
		Dim:           settings.EmbeddingDim,
		Available:     s.registry.Models(),
		GeminiDefault: "gemini-embedding-001",
	}
}

// ResolveEmbeddingTarget은 전환 대상을 검증만 수행한다(실제 전환 전 400 응답용).
func (s *Service) ResolveEmbeddingTarget(providerName, model string, dim *int) (domain.Target, error) {
	target, err := s.registry.Resolve(providerName, model, dim)
	if err != nil {
		return domain.Target{}, domain.BadRequest(err.Error())
	}
	return target, nil
}

// SwitchEmbeddingModel은 .env 갱신 → 설정·구현체 교체 → 전량 재임베딩을 수행한다.
func (s *Service) SwitchEmbeddingModel(ctx context.Context, providerName, model string, dim *int) (domain.SwitchResult, error) {
	var result domain.SwitchResult

	target, err := s.registry.Resolve(providerName, model, dim)
	if err != nil {
		return result, domain.BadRequest(err.Error())
	}
	settings := s.cfg.Get()

	// Gemini 전환 시 API 키 사전 확인 (환경변수 또는 기존 .env)
	if providerName == "gemini" && os.Getenv("GOOGLE_API_KEY") == "" &&
		!envfile.HasKey(settings.EnvFile, "GOOGLE_API_KEY") {
		return result, domain.BadRequest("GOOGLE_API_KEY가 없습니다. .env에 먼저 설정하세요")
	}

	// .env 갱신(주석 보존 + 백업) 후 프로세스 설정도 즉시 동기화
	updates := map[string]string{
		"EMBEDDING_PROVIDER":     target.Provider,
		"EMBEDDING_DIM":          strconv.Itoa(target.Dim),
		"EMBEDDING_MODEL":        settings.EmbeddingModel,
		"GEMINI_EMBEDDING_MODEL": settings.GeminiEmbeddingModel,
	}
	if target.Provider == "gemini" {
		updates["GEMINI_EMBEDDING_MODEL"] = target.Model
	} else {
		updates["EMBEDDING_MODEL"] = target.Model
	}
	backup, err := envfile.Update(settings.EnvFile, updates)
	if err != nil {
		return result, err
	}
	for key, value := range updates {
		if err := os.Setenv(key, value); err != nil {
			return result, err
		}
	}
	newSettings := settings
	newSettings.EmbeddingProvider = target.Provider
	newSettings.EmbeddingDim = target.Dim
	newSettings.EmbeddingModel = updates["EMBEDDING_MODEL"]
	newSettings.GeminiEmbeddingModel = updates["GEMINI_EMBEDDING_MODEL"]
	s.cfg.Set(newSettings)

	// 임베딩은 파생 데이터: 전량 삭제 후 새 구현체로 재임베딩
	newEmbedder, err := s.registry.Build(target.Provider, newSettings)
	if err != nil {
		return result, err
	}
	s.active.Swap(newEmbedder)

	if err := s.embeds.DeleteAll(ctx); err != nil {
		return result, err
	}
	reEmbedded, err := s.runner.EmbedAll(ctx)
	if err != nil {
		return result, err
	}

	result = domain.SwitchResult{
		Provider: target.Provider, Model: target.Model, Dim: target.Dim,
		ReEmbedded: reEmbedded, EnvBackup: backup,
	}
	slog.Info("임베딩 모델 전환 완료", "provider", result.Provider, "model", result.Model)
	return result, nil
}

// ---------- 캐시 ----------

// ListCache는 최근 분석 캐시 20건을 돌려준다.
func (s *Service) ListCache(ctx context.Context) ([]domain.CacheRow, error) {
	return s.cache.List(ctx, 20)
}

// ClearCache는 전체 캐시를 지우고 건수를 반환한다.
func (s *Service) ClearCache(ctx context.Context) (int64, error) {
	return s.cache.Clear(ctx)
}

// DeleteCache는 특정 (공고, 사용자) 캐시를 지운다.
func (s *Service) DeleteCache(ctx context.Context, jobID, userID int64) error {
	affected, err := s.cache.Delete(ctx, jobID, userID)
	if err != nil {
		return err
	}
	if affected == 0 {
		return domain.NotFound(fmt.Sprintf("캐시 없음: job=%d user=%d", jobID, userID))
	}
	return nil
}

// ---------- 백그라운드 트리거 ----------

// TriggerPipeline은 파이프라인을 백그라운드로 즉시 실행한다.
// 실행 중 파이프라인이 있거나(409) 가드에 걸리면(409) 오류다.
func (s *Service) TriggerPipeline(ctx context.Context) error {
	running, err := s.runs.HasRunning(ctx)
	if err != nil {
		return err
	}
	if running {
		return domain.Conflict("파이프라인이 이미 실행 중입니다")
	}
	if !s.guard.Start("pipeline", func(bg context.Context) error {
		slog.Info("어드민 요청으로 파이프라인 즉시 실행")
		_, err := s.runner.RunPipeline(bg)
		return err
	}) {
		return domain.Conflict("동일 작업이 이미 실행 중입니다")
	}
	return nil
}

// StartBackground는 이름 기반 가드로 임의 작업을 실행한다 (어드민 임베딩 작업용).
func (s *Service) StartBackground(name string, fn func(ctx context.Context) error) bool {
	return s.guard.Start(name, fn)
}

// RunningOps는 실행 중 작업 목록이다.
func (s *Service) RunningOps() []string { return s.guard.Running() }

func clamp(v, def int) int {
	if v < 1 {
		v = 1
	}
	if v > 100 {
		v = 100
	}
	return v
}
